library/Faktory/Pool.hs

```diff
@@ -0,0 +1,126 @@
+module Faktory.Pool
+  ( FaktoryPool
+  , HasFaktoryPool (..)
+
+    -- * Pool Construction
+  , Settings
+  , PoolSettings
+  , newFaktoryPool
+
+    -- * Pool use
+  , perform
+  , buildJob
+
+    -- * Direct access
+  , withProducer
+  , takeProducer
+
+    -- * Re-exports
+  , module Faktory.Job
+  ) where
+
+import Faktory.Prelude
+
+import Control.Monad.IO.Class (MonadIO (..))
+import Control.Monad.Reader (MonadReader, asks)
+import Data.Aeson (ToJSON)
+import Data.Pool (Pool)
+import qualified Data.Pool as Pool
+import Faktory.Job hiding (buildJob, perform)
+import qualified Faktory.Job as Job
+import Faktory.Producer
+import Faktory.Settings (PoolSettings (..), Settings)
+import GHC.Stack (HasCallStack)
+import Lens.Micro (Lens', (^.))
+import UnliftIO (MonadUnliftIO, withRunInIO)
+
+-- |
+--
+-- @since 1.1.3.0
+type FaktoryPool = Pool Producer
+
+-- |
+--
+-- @since 1.1.3.0
+class HasFaktoryPool env where
+  faktoryPoolL :: Lens' env FaktoryPool
+
+instance HasFaktoryPool FaktoryPool where
+  faktoryPoolL = id
+
+-- | Build a 'FaktoryPool' with the given settings
+--
+-- See 'Settings', 'envSettings', 'PoolSettings', and 'envPoolSettings'.
+--
+-- @since 1.1.3.0
+newFaktoryPool
+  :: MonadIO m
+  => Settings
+  -> PoolSettings
+  -> m FaktoryPool
+newFaktoryPool settings PoolSettings {..} = do
+  liftIO
+    . Pool.newPool
+    $ Pool.defaultPoolConfig
+      (newProducer settings)
+      closeProducer
+      (fromIntegral settingsTimeout)
+      (fromIntegral settingsSize)
+
+-- | 'Faktory.Job.perform' but using a 'Producer' from the pool
+--
+-- @since 1.1.3.0
+perform
+  :: ( MonadUnliftIO m
+     , MonadReader env m
+     , HasFaktoryPool env
+     , ToJSON arg
+     , HasCallStack
+     )
+  => JobOptions
+  -> arg
+  -> m JobId
+perform options arg = do
+  withProducer $ \producer -> do
+    liftIO $ Job.perform options producer arg
+
+-- | 'Faktory.Job.buildJob' but using a 'Producer' from the pool
+--
+-- @since 1.1.3.0
+buildJob
+  :: (MonadUnliftIO m, MonadReader env m, HasFaktoryPool env)
+  => JobOptions
+  -> arg
+  -> m (Job arg)
+buildJob options arg = do
+  withProducer $ \producer -> do
+    liftIO $ Job.buildJob options producer arg
+
+-- | Acquire a 'Producer', use it, and return it to the pool
+--
+-- @since 1.1.3.0
+withProducer
+  :: (MonadUnliftIO m, MonadReader env m, HasFaktoryPool env)
+  => (Producer -> m a)
+  -> m a
+withProducer f = do
+  p <- asks (^. faktoryPoolL)
+  withRunInIO $ \runInIO -> do
+    Pool.withResource p $ runInIO . f
+
+-- | Get a 'Producer' from the pool along with an action to return it
+--
+-- You should prefer 'withProducer' if at all possible. With this function you
+-- are responsible to ensure the return action is called (e.g. with 'finally').
+--
+-- This is only necessary if you are operating in a monad that doesn't have
+-- 'MonadUnliftIO' (like 'ConduitT'), so you need to take and return a
+-- 'Producer' separately (e.g. with 'bracketP').
+--
+-- @since 1.1.3.0
+takeProducer
```

Comment:
I'm onboard with exposing a "correct" `withProducer` function, but since `FaktoryPool` is a type synonym a client of this library can always write this "unsafe" one themselves. 

I guess the real question is do we ever expect to have an opaque `FaktoryPool` instead of a type synonym? If so then it seems like we should bite the `newtype` bullet now.

Reply:
Yeah, I was mostly just aping `SqlPool`. I tend to not really care about that particular safety. No one is going to "drop down" and be unsafe by accident. And if they have a reason, well then, I'm happy to ablige. But if you disagree I'll `newtype` it, I don't feel strongly.

I'm also silently holding out hope that some day we can figure out a way to stop doing this every single time, and instead be able to write code like,

```hs
foo
  :: ( MonadReader env m
     , HasPool env SqlBackend
     , HasPool env KafkaConnection
     , HasPool env Faktory.Producer
```

But I just don't know that that can ever work.

Reply:
I think I'm just a bit allergic to synonyms in general, but I don't feel strongly enough to want to see this changed.

Reply:
I did it.

Totally minor change, and the downstream code was unaffected. Might as well be safer :shrug: